[FSDP] Break up `_post_backward_hook` into smaller funcs

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107784
* #106080
* __->__ #106068

The post-backward hook has some complexity due to the different paths: {no communication hook, communication hook} x {`NO_SHARD`, `FULL_SHARD`/`SHARD_GRAD_OP`, `HYBRID_SHARD`/`_HYBRID_SHARD_ZERO2`} plus some options like CPU offloading and `use_orig_params=True` (requiring using sharded gradient views).

The PR following this one that adds async all-reduce for HSDP further complicates this since the bottom-half after all-reduce must still be run in the separate all-reduce stream, making it more unwieldy to unify with the existing bottom-half.

Nonetheless, this PR breaks up the post-backward hook into smaller logical functions to hopefully help readability.


Differential Revision: [D47852461](https://our.internmc.facebook.com/intern/diff/D47852461)

Commit:
[FSDP] Break up `_post_backward_hook` into smaller funcs

[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"




[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D47852461).

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"


Differential Revision: [D47852461](https://our.internmc.facebook.com/intern/diff/D47852461)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"


Differential Revision: [D47852461](https://our.internmc.facebook.com/intern/diff/D47852461)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"


The post-backward hook has some complexity due to the different paths: {no communication hook, communication hook} x {`NO_SHARD`, `FULL_SHARD`/`SHARD_GRAD_OP`, `HYBRID_SHARD`/`_HYBRID_SHARD_ZERO2`} plus some options like CPU offloading and `use_orig_params=True` (requiring using sharded gradient views).

The PR following this one that adds async all-reduce for HSDP further complicates this since the bottom-half after all-reduce must still be run in the separate all-reduce stream, making it more unwieldy to unify with the existing bottom-half.

Nonetheless, this PR breaks up the post-backward hook into smaller logical functions to hopefully help readability.


Differential Revision: [D47852461](https://our.internmc.facebook.com/intern/diff/D47852461)

[ghstack-poisoned]

ezyang commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -951,6 +833,154 @@ def _should_free_in_backward(
     )
 
 
```
Are you perhaps willing to add small top level comments for each of the new functions? :)

awgu replied:
I added the comments in the subsequent PR since the async HSDP refactors these a bit more now.

ezyang reviewed (approved):
Verified code motion only

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"


The post-backward hook has some complexity due to the different paths: {no communication hook, communication hook} x {`NO_SHARD`, `FULL_SHARD`/`SHARD_GRAD_OP`, `HYBRID_SHARD`/`_HYBRID_SHARD_ZERO2`} plus some options like CPU offloading and `use_orig_params=True` (requiring using sharded gradient views).

The PR following this one that adds async all-reduce for HSDP further complicates this since the bottom-half after all-reduce must still be run in the separate all-reduce stream, making it more unwieldy to unify with the existing bottom-half.

Nonetheless, this PR breaks up the post-backward hook into smaller logical functions to hopefully help readability.


Differential Revision: [D47852461](https://our.internmc.facebook.com/intern/diff/D47852461)

[ghstack-poisoned]

Commit:
Update on "[FSDP] Break up `_post_backward_hook` into smaller funcs"


The post-backward hook has some complexity due to the different paths: {no communication hook, communication hook} x {`NO_SHARD`, `FULL_SHARD`/`SHARD_GRAD_OP`, `HYBRID_SHARD`/`_HYBRID_SHARD_ZERO2`} plus some options like CPU offloading and `use_orig_params=True` (requiring using sharded gradient views).

The PR following this one that adds async all-reduce for HSDP further complicates this since the bottom-half after all-reduce must still be run in the separate all-reduce stream, making it more unwieldy to unify with the existing bottom-half.

Nonetheless, this PR breaks up the post-backward hook into smaller logical functions to hopefully help readability.


Differential Revision: [D47852461](https://our.internmc.facebook.com/intern/diff/D47852461)

[ghstack-poisoned]
